hydroflow/src/builder/hydroflow_builder.rs

```diff
@@ -0,0 +1,124 @@
+use super::build::{PullBuild, PushBuild};
+use super::connect::{PullConnect, PushConnect};
+use super::surface::pivot::PivotSurface;
+
+use std::rc::Rc;
+use std::sync::mpsc::SyncSender;
+
+use crate::compiled::pivot::Pivot;
+use crate::scheduled::graph::Hydroflow;
+use crate::scheduled::graph_ext::GraphExt;
+use crate::scheduled::handoff::{CanReceive, Handoff};
+use crate::scheduled::input::Input;
+
+use super::surface::pull_handoff::HandoffPullSurface;
+use super::surface::push_handoff::HandoffPushSurfaceReversed;
+use super::surface::push_start::StartPushSurface;
+use super::surface::{PullSurface, PushSurfaceReversed};
+
+/// The user-facing entry point for the Surface API.
+#[derive(Default)]
+pub struct HydroflowBuilder {
+    hydroflow: Hydroflow,
+    // TODO(mingwei): use a dedicated trait instead of FnOnce?
+    handoff_connectors: Vec<Box<dyn FnOnce(&mut Hydroflow)>>,
+}
+impl HydroflowBuilder {
+    /// Creates a handoff, returning push and pull ends which can be chained
+    /// using the Surface API.
+    pub fn make_handoff<H, T>(
+        &mut self,
+    ) -> (HandoffPushSurfaceReversed<H, T>, HandoffPullSurface<H>)
+    where
+        H: Handoff + CanReceive<T>,
+    {
+        let push_port = Default::default();
+        let push = HandoffPushSurfaceReversed::new(Rc::clone(&push_port));
+
+        let pull_port = Default::default();
+        let pull = HandoffPullSurface::new(Rc::clone(&pull_port));
+
+        self.handoff_connectors.push(Box::new(move |hydroflow| {
+            if let (Some(output_port), Some(input_port)) = (push_port.take(), pull_port.take()) {
+                hydroflow.add_edge(output_port, input_port);
+            } else {
+                panic!("Handoff was never connected!!"); // TODO(mingwei): more informative error messages.
+            }
+        }));
+        (push, pull)
+    }
+
+    /// Adds a `pivot` created via the Surface API.
+    pub fn add_subgraph<Pull, Push>(&mut self, pivot: PivotSurface<Pull, Push>)
+    where
+        Pull: 'static + PullSurface,
+        Push: 'static + PushSurfaceReversed<ItemIn = Pull::ItemOut>,
+    {
+        let ((pull_connect, push_connect), (mut pull_build, mut push_build)) = pivot.into_parts();
+
+        let (input_ports, output_ports) = self
+            .hydroflow
+            .add_subgraph::<_, Pull::InputHandoffs, Push::OutputHandoffs>(
+                move |_ctx, recv_ctx, send_ctx| {
+                    let pull = pull_build.build(recv_ctx);
```

Comment:
Is it possible to build pull sides that output to a handoff? I think there's cases where that can (in theory, at least) work structurally:
```
                       ...
                        |
                        v
Input handoff -> X -> Pivot -> ...
                 |
                 v
         Output Handoff
```
If that complicates things I don't think it's a huge deal if we don't support it, but it would be nice to have the flexibility.

Reply:
I think it is possible, it would look like a pull operator which takes in a `PushSurfaceReversed` (same type that tee takes in, but only one instead of two)